x/oob/writer.go

```diff
@@ -0,0 +1,118 @@
+package oob
+
+import (
+	"fmt"
+	"io"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/Jigsaw-Code/outline-sdk/x/sockopt"
+)
+
+type oobWriter struct {
+	conn        *net.TCPConn
+	opts        sockopt.TCPOptions
+	resetTTL    sync.Once
+	setTTL      sync.Once
+	oobPosition int64
+	sd          SocketDescriptor
+	oobByte     byte // Byte to send as OOB
+	disOOB      bool // Flag to enable disOOB mode
+	delay       time.Duration
+}
+
+var _ io.Writer = (*oobWriter)(nil)
+
+type oobWriterReaderFrom struct {
+	*oobWriter
+	rf io.ReaderFrom
+}
+
+// NewWriter creates an [io.Writer] that sends an OOB byte at the specified "oobPosition".
+// If disOOB is enabled, it will apply the --disOOB strategy.
+// "oobByte" specifies the value of the byte to send out-of-band.
+func NewWriter(
+	conn *net.TCPConn,
+	opts sockopt.TCPOptions,
+	oobPosition int64,
+	oobByte byte,
+	disOOB bool,
+	delay time.Duration,
+) io.Writer {
+	return &oobWriter{conn: conn, opts: opts, oobPosition: oobPosition, oobByte: oobByte, disOOB: disOOB, delay: delay}
+}
+
+func (w *oobWriter) Write(data []byte) (int, error) {
+	var written int
+	var err error
+
+	if w.oobPosition > 0 && w.oobPosition < int64(len(data))-1 {
```

Comment:
There's a corner case here. Probably worth capturing in the tests.

If the oob position aligns with the end of a write, it will never happen.

The problem is that you are dependent on an extra byte from the following write to push the OOB byte.

There's a way to fix that without requiring memory allocation. The SendMsg call can take the OOB as a separate buffer!

On Windows, the [WSASend](https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-wsasend) call takes a list of buffers, so just append the OOB to that list.